### PR TITLE
SFD-2281 Service Data ETL

### PR DESCRIPTION
**Pull request by gwaardenburg**

### CODE REVIEW

**Review by ArifNhs — DISMISSED** [review]
Approved

File: build/automation/var/platform-texas/v1/account-live-k8s-prod.mk [hunk 1]
```diff
@@ -5,6 +5,16 @@ AWS_WAF_ACL_ID = ff530a4e-689d-4d90-b3ab-ac2160b5863d
 
 TF_VAR_terraform_platform_state_store = nhsd-texasplatform-terraform-state-store-lk8s-$(AWS_ACCOUNT_NAME)
```
Comment (jope1) [inline build/automation/var/platform-texas/v1/account-live-k8s-prod.mk:7]: For prod, vars are:
TF_VAR_dos_replica_db := core-dos-sf-replica.dos-db-rds
TF_VAR_dos_security_group := live-lk8s-prod-uec-sf-core-dos-sf-replica-sg
TF_VAR_dos_read_replica_secret_name := core-dos/deployment
TF_VAR_ccg_etl_source_db := pathwaysdos

File: build/automation/var/profile/dev.mk [hunk 2]
```diff
@@ -44,7 +44,7 @@ NAME_PUBLIC_PRIORITY := 4
 # Infrastructure variables
 
 DEPLOYMENT_STACKS = application
-INFRASTRUCTURE_STACKS = elasticsearch,authentication,roles
+INFRASTRUCTURE_STACKS = elasticsearch,authentication,roles,dos_replcia_etl
```
Comment (jope1) [inline build/automation/var/profile/dev.mk:47]: Can we call the stack services_etl please?

File: infrastructure/stacks/dos_replica_etl/data.tf [hunk 3]
```diff
@@ -0,0 +1,35 @@
+# Lambda
+data "archive_file" "dos_replica_etl_function" {
```
Comment (jope1) [inline infrastructure/stacks/dos_replica_etl/data.tf:2]: Lets do a general find and replace of 'dos-replica' to 'services'

File: infrastructure/stacks/dos_replica_etl/data.tf [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+data "terraform_remote_state" "vpc" {
+  backend = "s3"
+  config = {
+    bucket = var.terraform_platform_state_store
+    key    = var.vpc_terraform_state_key
+    region = var.aws_region
+  }
+}
+
+data "aws_security_group" "dos_application_security_group" {
```
Comment (jope1) [inline infrastructure/stacks/dos_replica_etl/data.tf:17]: This is the service_finder_replica_sg can we change the name to reflect please?

File: infrastructure/stacks/dos_replica_etl/data.tf [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  name = var.dos_security_group
```
Comment (jope1) [inline infrastructure/stacks/dos_replica_etl/data.tf:18]: Lets make this name clearer too please

File: infrastructure/stacks/dos_replica_etl/data.tf [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+data "aws_security_group" "elasticsearch_security_group" {
+  name = "${var.service_prefix}-elastic-search"
+}
+
+data "aws_lambda_layer_version" "dos_python_libs" {
+  layer_name = var.core_dos_python_libs
+}
+
+data "aws_secretsmanager_secret" "dos_read_replica_secret_name" {
```
Comment (jope1) [inline infrastructure/stacks/dos_replica_etl/data.tf:29]: dos_sf_read_replica_secret_name

File: infrastructure/stacks/dos_replica_etl/dos_replica_etl.tf [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+  source_code_hash = data.archive_file.dos_replica_etl_function.output_base64sha256
+  runtime          = local.dos_replica_etl_runtime
+  timeout          = local.dos_replica_etl_timeout
+  memory_size      = local.dos_replica_etl_memory_size
+  publish          = true
+  tags             = local.standard_tags
+  layers           = [local.dos_replica_etl_core_dos_python_libs_arn]
+  environment {
+    variables = {
+      USR                = local.dos_replica_etl_db_user
```
Comment (jope1) [inline infrastructure/stacks/dos_replica_etl/dos_replica_etl.tf:16]: Lets replace dos_replica with dos_sf_replica

File: infrastructure/stacks/dos_replica_etl/dos_replica_etl.tf [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+    }
+  ]
+}
+
+resource "aws_security_group_rule" "dos_replica_etl_security_group_rule" {
+  type                     = "ingress"
+  from_port                = 443
+  to_port                  = 443
+  protocol                 = "tcp"
+  description              = "allows lambda etl process to insert data into ES"
```
Comment (jope1) [inline infrastructure/stacks/dos_replica_etl/dos_replica_etl.tf:64]: Update description:
"Allows SF Services ETL Lambda to access SF Elasticsearch datastore"

File: infrastructure/stacks/dos_replica_etl/variables.tf [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+variable "vpc_terraform_state_key" { description = "The VPC key in the terraform state bucket" }
+
+variable "es_domain_name" { description = "Elastic search domain name" }
+
+variable "service_prefix" { description = "The prefix used to adhere to the naming conventions" }
+
+# === Stack Specific ============================================================
+
+variable "dos_replica_db" { description = "The DOS read replica to get the postcodes from" }
```
Comment (jope1) [inline infrastructure/stacks/dos_replica_etl/variables.tf:20]: dos_sf_replica_db

File: infrastructure/stacks/dos_replica_etl/variables.tf [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+
+variable "dos_security_group" { description = "The security group for the dos read replica" }
```
Comment (jope1) [inline infrastructure/stacks/dos_replica_etl/variables.tf:22]: dos_sf_replica_sg

File: infrastructure/stacks/dos_replica_etl/variables.tf [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+
+variable "dos_read_replica_secret_name" { description = "The dos read replica secret name" }
+
+variable "dos_read_replica_secret_key" { description = "The dos read replica secret key" }
+
+variable "dos_replica_etl_db_user" { description = "The dos read replica user name" }
```
Comment (jope1) [inline infrastructure/stacks/dos_replica_etl/variables.tf:28]: services_etl_db_user

File: infrastructure/stacks/dos_replica_etl/variables.tf [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+
+variable "dos_replica_etl_source_db" { description = "The dos read replica source database" }
```
Comment (jope1) [inline infrastructure/stacks/dos_replica_etl/variables.tf:30]: services_etl_source_db

File: infrastructure/stacks/dos_replica_etl/locals.tf [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+
+  dos_replica_etl_function_name = "uec-sf-${var.profile}-dos-replica-etl"
+  dos_replica_etl_description   = "Service Finder function to extract data out of the DoS Read Replica and insert it into the elasticsearch datastore. Runs every 5 mins"
+  dos_replica_etl_runtime       = "python3.8"
+  //This is set to 3 mins and 59 seconds as the timer is set to run every 4 mins so it will timeout before the next job starts
+  dos_replica_etl_timeout     = 239
+  dos_replica_etl_memory_size = 10240
+
+  dos_replica_etl_core_dos_python_libs_arn = data.aws_lambda_layer_version.dos_python_libs.arn
+  dos_replica_etl_vpc_security_group       = data.aws_security_group.dos_application_security_group.id
```
Comment (jope1) [inline infrastructure/stacks/dos_replica_etl/locals.tf:11]: dos_sf_replica_db_sg

**Review by jope1 — CHANGES_REQUESTED** [review]
Do we need all of the stuff under the stacks/dos-replica-etl/functions directory? Don't we just need the functions.zip file?

**Review by jope1 — APPROVED** [review]

**Pull request merged**